src/creatures/players/player.cpp

```diff
@@ -7068,6 +7077,766 @@ void Player::dismount() {
+}
+
+void Player::setCurrentShader(uint16_t shaderId) {
+	currentShader = shaderId;
+}
+
+bool Player::toggleShader(bool shader) {
+	if ((OTSYS_TIME() - lastToggleShader) < 3000 && !wasShadered) {
+		sendCancelMessage(RETURNVALUE_YOUAREEXHAUSTED);
+		return false;
+	}
+
+	if (shader) {
+		if (isShadered()) {
+			return false;
+		}
+
+		const auto &playerOutfit = Outfits::getInstance().getOutfitByLookType(getPlayer(), defaultOutfit.lookType);
+		if (!playerOutfit) {
+			return false;
+		}
+
+		uint16_t currentShaderId = getCurrentShader();
+		if (currentShaderId == 0) {
+			sendOutfitWindow();
+			return false;
+		}
+
+		auto currentShaderPtr = g_game().attachedeffects->getShaderByID(currentShaderId);
+		if (!currentShaderPtr) {
+			return false;
+		}
+
+		Shader* currentShader = currentShaderPtr.get();
+
+		if (!hasShader(currentShader)) {
+			setCurrentShader(0);
+			sendOutfitWindow();
+			return false;
+		}
+
+		if (hasCondition(CONDITION_OUTFIT)) {
+			sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
+			return false;
+		}
+
+		defaultOutfit.lookShader = currentShader->id;
+
+	} else {
+		if (!isShadered()) {
+			return false;
+		}
+
+		disshader();
+	}
+
+	g_game().internalCreatureChangeOutfit(static_self_cast<Player>(), defaultOutfit);
+	lastToggleShader = OTSYS_TIME();
+	return true;
+}
+
+bool Player::tameShader(uint16_t shaderId) {
+	auto shaderPtr = g_game().attachedeffects->getShaderByID(shaderId);
+	if (!shaderPtr) {
+		return false;
+	}
+
+	if (hasShader(shaderPtr.get())) {
+		return false;
+	}
+
+	shaders.insert(shaderId);
+	return true;
+}
+
+bool Player::untameShader(uint16_t shaderId) {
+	auto shaderPtr = g_game().attachedeffects->getShaderByID(shaderId);
+	if (!shaderPtr) {
+		return false;
+	}
+
+	if (!hasShader(shaderPtr.get())) {
+		return false;
+	}
+
+	shaders.erase(shaderId);
+
+	if (getCurrentShader() == shaderId) {
+		if (isShadered()) {
+			disshader();
+			g_game().internalCreatureChangeOutfit(static_self_cast<Player>(), defaultOutfit);
+		}
+
+		setCurrentShader(0);
+	}
+
+	return true;
+}
+
+bool Player::hasShader(const Shader* shader) const {
+	if (isAccessPlayer()) {
+		return true;
+	}
+
+	return shaders.find(shader->id) != shaders.end();
+}
+
+bool Player::hasShaders() const {
+	for (const auto &shader : g_game().attachedeffects->getShaders()) {
+		if (hasShader(shader.get())) {
+			return true;
+		}
+	}
+	return false;
+}
+
+void Player::disshader() {
+	defaultOutfit.lookShader = 0;
+}
+
+std::string Player::getCurrentShader_NAME() const {
+	uint16_t currentShaderId = getCurrentShader();
+	const auto &currentShader = g_game().attachedeffects->getShaderByID(static_cast<uint8_t>(currentShaderId));
+
+	if (currentShader != nullptr) {
+		return currentShader->name;
+	} else {
+		return "Outfit - Default";
+	}
+}
+
+bool Player::addCustomOutfit(const std::string &type, const std::variant<uint16_t, std::string> &idOrName) {
+	// test proposal
+
+	uint16_t elementId;
+	if (std::holds_alternative<uint16_t>(idOrName)) {
```

Comment:
If this is used very frequently or continuously, it is advisable to avoid using holds alternative, it adds significant overhead as it needs to evaluate the value at runtime.